packages/astro/src/assets/fonts/vite-plugin-fonts.ts

```diff
@@ -0,0 +1,237 @@
+		// TODO: remove once fonts are stabilized
+		return {
+			name: 'astro:fonts:fallback',
+			config() {
+				return {
+					build: {
+						rollupOptions: {
+							external: [VIRTUAL_MODULE_ID],
+						},
+					},
+				};
+			},
+		};
+	}
+
+	const providers: Array<FontProvider<string>> = settings.config.experimental.fonts.providers ?? [];
+	const families: Array<FontFamily<string>> = settings.config.experimental.fonts.families;
+
+	const baseUrl = removeTrailingForwardSlash(settings.config.base) + URL_PREFIX;
+
+	let resolvedMap: Map<string, { preloadData: PreloadData; css: string }> | null = null;
+	/** Key is `${hash}.${ext}`, value is a URL */
+	let collected: Map<string, string> | null = null;
+	let isBuild: boolean;
+	let cache: Cache['cache'] | null = null;
+
+	return {
+		name: 'astro:fonts',
+		config(_, { command }) {
+			isBuild = command === 'build';
+		},
+		async buildStart() {
+			const { h64ToString } = await xxhash();
+
+			const resolved = await resolveProviders({
+				settings,
+				providers,
+			});
+
+			const storage = createStorage({
+				// In build, first check the build cache then fallback to dev cache
+				// In dev, only check dev cache
+				bases: [
+					new URL(CACHE_DIR, settings.dotAstroDir),
+					...(isBuild ? [new URL(CACHE_DIR, settings.config.cacheDir)] : []),
+				],
+			});
+			cache = createCache({ storage }).cache;
+
+			const { resolveFont } = await unifont.createUnifont(
+				resolved.map((e) => e.provider(e.config)),
+				{ storage },
+			);
+			resolvedMap = new Map();
+			collected = new Map();
+			for (const family of families) {
+				const resolvedOptions: unifont.ResolveFontOptions = {
+					weights: family.weights ?? DEFAULTS.weights,
+					styles: family.styles ?? DEFAULTS.styles,
+					subsets: family.subsets ?? DEFAULTS.subsets,
+					fallbacks: family.fallbacks ?? DEFAULTS.fallbacks,
+				};
+				// TODO: https://github.com/unjs/unifont/issues/108
+				const { fonts: fontsData, fallbacks } = await resolveFont(family.name, resolvedOptions, [
+					family.provider,
+				]);
+
+				// TODO: use fontaine if needed
+				const preloadData: PreloadData = [];
+				let css = '';
+				for (const data of fontsData) {
+					for (const source of data.src as unknown as Array<Record<string, string>>) {
+						const key = 'name' in source ? 'name' : 'url';
+						const hash = h64ToString(source[key]) + extname(source[key]);
+						const url = baseUrl + hash;
+						if (!collected.has(hash)) {
+							collected.set(hash, source[key]);
+							preloadData.push({ url, type: hash.split('.')[1] });
+						}
+						source[key] = url;
+					}
+					css += generateFontFace(family.name, data);
+				}
+				resolvedMap.set(family.name, { preloadData, css });
+			}
+		},
+		async configureServer(server) {
+			const logManager = createLogManager(logger);
+			// Base is taken into account by default
+			server.middlewares.use(URL_PREFIX, async (req, res, next) => {
```

Comment:
If `trailingSlash` is set to `"never"`, the endpoint `/_fonts/` should never hit